File: projects/state-machine/python/src/transition.py
"""
Transition definitions for the state machine framework.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional, Union

from .action import Action, FunctionAction, TransitionAction
from .event import Event
from .guard import Guard, FunctionGuard, always_true
from .state import State


@dataclass
class Transition:
    """
    Represents a transition between two states triggered by an event.

    A transition defines:
    - Source state (from_state)
    - Target state (to_state)
    - Triggering event
    - Optional guard condition
    - Optional action to execute

    Example:
        >>> transition = Transition(
        ...     from_state=State("Off"),
        ...     to_state=State("On"),
        ...     event=Event("TurnOn"),
        ...     guard=FunctionGuard(lambda f, t, e, c: True),
        ...     action=FunctionAction(lambda f, t, e, c: print("Turned on!")),
        ...     description="Turn the light on"
        ... )
    """

    from_state: State
    to_state: State
    event: Event
    guard: Optional[Guard] = field(default_factory=always_true)
    action: Optional[Action] = None
    description: Optional[str] = None

    def can_transition(
        self,
        context: Optional[Dict[str, Any]] = None,
    ) -> bool:
        """
        Check if this transition can occur.

        Args:
            context: Optional context data for guard evaluation

        Returns:
            True if the guard condition is satisfied
        """
        if self.guard is None:
            return True
        return self.guard.check(self.from_state, self.to_state, self.event, context)

    def execute(
        self,
        context: Optional[Dict[str, Any]] = None,
    ) -> Optional[Any]:
        """
        Execute the transition action.

        Args:
            context: Optional context data for the action

        Returns:
            Result of the action, if any
        """
        if self.action is None:
            return None
        return self.action.execute(self.from_state, self.to_state, self.event, context)

    def matches(self, from_state: State, event: Event) -> bool:
        """
        Check if this transition matches a source state and event.

        Args:
            from_state: The source state to match
            event: The event to match

        Returns:
            True if this transition matches
        """
        return self.from_state == from_state and self.event == event

    def __repr__(self) -> str:
        parts = [f"{self.from_state} -> {self.to_state} on {self.event}"]
        if self.guard and not isinstance(self.guard, type(always_true())):
            parts.append(f"when {self.guard}")
        if self.action:
            parts.append(f"do {self.action}")
        if self.description:
            parts.append(f"# {self.description}")
        return "Transition(" + ", ".join(parts) + ")"

    def __str__(self) -> str:
        if self.description:
            return self.description
        return f"{self.from_state} -> {self.to_state} on {self.event}"


class TransitionBuilder:
    """
    Builder pattern for creating transitions.

    Provides a fluent API for constructing transitions step by step.

    Example:
        >>> transition = (TransitionBuilder()
        ...     .from_state(State("Off"))
        ...     .to_state(State("On"))
        ...     .on(Event("TurnOn"))
        ...     .when(lambda f, t, e, c: True)
        ...     .do(lambda f, t, e, c: print("Turning on"))
        ...     .describe("Turn the light on")
        ...     .build())
    """

    def __init__(self) -> None:
        """Initialize the builder."""
        self._from_state: Optional[State] = None
        self._to_state: Optional[State] = None
        self._event: Optional[Event] = None
        self._guard: Optional[Guard] = None
        self._action: Optional[Action] = None
        self._description: Optional[str] = None

    def from_state(self, state: State) -> TransitionBuilder:
        """
        Set the source state.

        Args:
            state: The source state

        Returns:
            Self for chaining
        """
        self._from_state = state
        return self

    def to_state(self, state: State) -> TransitionBuilder:
        """
        Set the target state.

        Args:
            state: The target state

        Returns:
            Self for chaining
        """
        self._to_state = state
        return self

    def to(self, state: State) -> TransitionBuilder:
        """
        Alias for to_state.

        Args:
            state: The target state

        Returns:
            Self for chaining
        """
        return self.to_state(state)

    def on(self, event: Event) -> TransitionBuilder:
        """
        Set the triggering event.

        Args:
            event: The triggering event

        Returns:
            Self for chaining
        """
        self._event = event
        return self

    def when(self, guard: Union[Guard, Callable]) -> TransitionBuilder:
        """
        Set the guard condition.

        Args:
            guard: A Guard instance or callable

        Returns:
            Self for chaining
        """
        if callable(guard) and not isinstance(guard, Guard):
            guard = FunctionGuard(guard)
        self._guard = guard
        return self

    def do(self, action: Union[Action, Callable]) -> TransitionBuilder:
        """
        Set the transition action.

        Args:
            action: An Action instance or callable

        Returns:
            Self for chaining
        """
        if callable(action) and not isinstance(action, Action):
            action = FunctionAction(action)
        self._action = action
        return self

    def describe(self, description: str) -> TransitionBuilder:
        """
        Set the description.

        Args:
            description: Human-readable description

        Returns:
            Self for chaining
        """
        self._description = description
        return self

    def build(self) -> Transition:
        """
        Build the transition.

        Returns:
            The constructed Transition

        Raises:
            ValueError: If required fields are missing
        """
        if self._from_state is None:
            raise ValueError("Source state is required")
        if self._to_state is None:
            raise ValueError("Target state is required")
        if self._event is None:
            raise ValueError("Event is required")

        return Transition(
            from_state=self._from_state,
            to_state=self._to_state,
            event=self._event,
            guard=self._guard,
            action=self._action,
            description=self._description,
        )


def create_transition(
    from_state: State,
    to_state: State,
    event: Event,
    guard: Optional[Union[Guard, Callable]] = None,
    action: Optional[Union[Action, Callable]] = None,
    description: Optional[str] = None,
) -> Transition:
    """
    Factory function to create a transition.

    Args:
        from_state: Source state
        to_state: Target state
        event: Triggering event
        guard: Optional guard condition
        action: Optional action to execute
        description: Optional description

    Returns:
        A new Transition instance

    Example:
        >>> transition = create_transition(
        ...     State("Off"), State("On"), Event("TurnOn"),
        ...     guard=lambda f, t, e, c: True,
        ...     action=lambda f, t, e, c: print("On!"),
        ... )
    """
    if callable(guard) and not isinstance(guard, Guard):
        guard = FunctionGuard(guard)
    if callable(action) and not isinstance(action, Action):
        action = FunctionAction(action)

    return Transition(
        from_state=from_state,
        to_state=to_state,
        event=event,
        guard=guard,
        action=action,
        description=description,
    )

File: projects/state-machine/src/transition.rs
use crate::{Event, Result, State, StateMachineError};
use std::fmt::Debug;

/// Guard function type for transition conditions
pub type GuardFn<S, E> = Box<dyn Fn(&S, &S, &E) -> bool + Send + Sync>;

/// Action function type for transition side effects
pub type ActionFn<S, E> = Box<dyn Fn(&S, &S, &E) -> Result<()> + Send + Sync>;

/// Represents a transition between states
pub struct Transition<S: State, E: Event> {
    /// Source state
    pub from: S,
    /// Target state
    pub to: S,
    /// Event that triggers this transition
    pub event: E,
    /// Optional guard condition
    pub guard: Option<GuardFn<S, E>>,
    /// Optional action to execute on transition
    pub action: Option<ActionFn<S, E>>,
    /// Description of this transition
    pub description: Option<String>,
}

impl<S: State, E: Event> Transition<S, E> {
    /// Create a new transition without guard or action
    pub fn new(from: S, to: S, event: E) -> Self {
        Self {
            from,
            to,
            event,
            guard: None,
            action: None,
            description: None,
        }
    }

    /// Set a guard condition for this transition
    pub fn with_guard(mut self, guard: impl Fn(&S, &S, &E) -> bool + Send + Sync + 'static) -> Self {
        self.guard = Some(Box::new(guard));
        self
    }

    /// Set an action to execute when this transition occurs
    pub fn with_action(mut self, action: impl Fn(&S, &S, &E) -> Result<()> + Send + Sync + 'static) -> Self {
        self.action = Some(Box::new(action));
        self
    }

    /// Set a description for this transition
    pub fn with_description(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Check if the guard condition allows this transition
    pub fn check_guard(&self, event: &E) -> bool {
        match &self.guard {
            Some(guard) => guard(&self.from, &self.to, event),
            None => true,
        }
    }

    /// Execute the action for this transition
    pub fn execute_action(&self, event: &E) -> Result<()> {
        match &self.action {
            Some(action) => action(&self.from, &self.to, event),
            None => Ok(()),
        }
    }

    /// Check if this transition matches the given state and event
    pub fn matches(&self, state: &S, event: &E) -> bool {
        self.from == *state && self.event == *event
    }
}

impl<S: State, E: Event> Debug for Transition<S, E> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Transition")
            .field("from", &self.from)
            .field("to", &self.to)
            .field("event", &self.event)
            .field("has_guard", &self.guard.is_some())
            .field("has_action", &self.action.is_some())
            .field("description", &self.description)
            .finish()
    }
}

/// Builder for creating transitions
pub struct TransitionBuilder<S: State, E: Event> {
    from: Option<S>,
    to: Option<S>,
    event: Option<E>,
    guard: Option<GuardFn<S, E>>,
    action: Option<ActionFn<S, E>>,
    description: Option<String>,
}

impl<S: State, E: Event> TransitionBuilder<S, E> {
    /// Create a new transition builder
    pub fn new() -> Self {
        Self {
            from: None,
            to: None,
            event: None,
            guard: None,
            action: None,
            description: None,
        }
    }

    /// Set the source state
    pub fn from(mut self, from: S) -> Self {
        self.from = Some(from);
        self
    }

    /// Set the target state
    pub fn to(mut self, to: S) -> Self {
        self.to = Some(to);
        self
    }

    /// Set the trigger event
    pub fn on(mut self, event: E) -> Self {
        self.event = Some(event);
        self
    }

    /// Set a guard condition
    pub fn when(mut self, guard: impl Fn(&S, &S, &E) -> bool + Send + Sync + 'static) -> Self {
        self.guard = Some(Box::new(guard));
        self
    }

    /// Set an action to execute
    pub fn then(mut self, action: impl Fn(&S, &S, &E) -> Result<()> + Send + Sync + 'static) -> Self {
        self.action = Some(Box::new(action));
        self
    }

    /// Set a description
    pub fn describe(mut self, description: impl Into<String>) -> Self {
        self.description = Some(description.into());
        self
    }

    /// Build the transition
    pub fn build(self) -> Result<Transition<S, E>> {
        let from = self.from.ok_or_else(|| {
            StateMachineError::generic("Source state is required")
        })?;
        let to = self.to.ok_or_else(|| {
            StateMachineError::generic("Target state is required")
        })?;
        let event = self.event.ok_or_else(|| {
            StateMachineError::generic("Event is required")
        })?;

        Ok(Transition {
            from,
            to,
            event,
            guard: self.guard,
            action: self.action,
            description: self.description,
        })
    }
}

impl<S: State, E: Event> Default for TransitionBuilder<S, E> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestState {
        A,
        B,
        C,
    }

    #[derive(Debug, Clone, PartialEq, Eq, Hash)]
    enum TestEvent {
        GoToB,
        GoToC,
        GoToA,
    }

    impl State for TestState {}
    impl Event for TestEvent {}

    #[test]
    fn test_transition_new() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB);
        assert_eq!(t.from, TestState::A);
        assert_eq!(t.to, TestState::B);
        assert_eq!(t.event, TestEvent::GoToB);
        assert!(t.guard.is_none());
        assert!(t.action.is_none());
    }

    #[test]
    fn test_transition_matches() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB);
        assert!(t.matches(&TestState::A, &TestEvent::GoToB));
        assert!(!t.matches(&TestState::B, &TestEvent::GoToB));
        assert!(!t.matches(&TestState::A, &TestEvent::GoToC));
    }

    #[test]
    fn test_transition_guard_always_true() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB);
        assert!(t.check_guard(&TestEvent::GoToB));
    }

    #[test]
    fn test_transition_guard_with_condition() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB)
            .with_guard(|_from, _to, event| *event == TestEvent::GoToB);

        assert!(t.check_guard(&TestEvent::GoToB));
        assert!(!t.check_guard(&TestEvent::GoToC));
    }

    #[test]
    fn test_transition_action() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB)
            .with_action(|_from, _to, _event| {
                // Action executed successfully
                Ok(())
            });

        assert!(t.execute_action(&TestEvent::GoToB).is_ok());
    }

    #[test]
    fn test_transition_with_description() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB)
            .with_description("Move from A to B");

        assert_eq!(t.description, Some("Move from A to B".to_string()));
    }

    #[test]
    fn test_transition_builder_success() {
        let t = TransitionBuilder::new()
            .from(TestState::A)
            .to(TestState::B)
            .on(TestEvent::GoToB)
            .describe("A to B")
            .build()
            .unwrap();

        assert_eq!(t.from, TestState::A);
        assert_eq!(t.to, TestState::B);
        assert_eq!(t.event, TestEvent::GoToB);
        assert_eq!(t.description, Some("A to B".to_string()));
    }

    #[test]
    fn test_transition_builder_missing_from() {
        let result = TransitionBuilder::new()
            .to(TestState::B)
            .on(TestEvent::GoToB)
            .build();

        assert!(result.is_err());
    }

    #[test]
    fn test_transition_builder_missing_to() {
        let result = TransitionBuilder::new()
            .from(TestState::A)
            .on(TestEvent::GoToB)
            .build();

        assert!(result.is_err());
    }

    #[test]
    fn test_transition_builder_missing_event() {
        let result: crate::Result<Transition<TestState, TestEvent>> = TransitionBuilder::new()
            .from(TestState::A)
            .to(TestState::B)
            .build();

        assert!(result.is_err());
    }

    #[test]
    fn test_transition_debug() {
        let t = Transition::new(TestState::A, TestState::B, TestEvent::GoToB);
        let debug_str = format!("{:?}", t);
        assert!(debug_str.contains("A"));
        assert!(debug_str.contains("B"));
        assert!(debug_str.contains("GoToB"));
    }
}
